cli/src/cli.rs

```diff
@@ -1342,94 +1597,248 @@ fn do_process_deploy(
+    // Create and add final message
+    let final_message = if loader_id == bpf_loader_upgradeable::id() {
+        Message::new(
+            &bpf_loader_upgradeable::deploy_with_max_program_len(
+                &config.signers[0].pubkey(),
+                &program_signer.pubkey(),
+                &buffer_signer.pubkey(),
+                upgrade_authority.as_ref(),
+                rpc_client.get_minimum_balance_for_rent_exemption(
+                    UpgradeableLoaderState::program_len()?,
+                )?,
+                data_len,
+            )?,
+            Some(&config.signers[0].pubkey()),
+        )
+    } else {
+        Message::new(
+            &[loader_instruction::finalize(
+                &buffer_signer.pubkey(),
+                &loader_id,
+            )],
+            Some(&config.signers[0].pubkey()),
+        )
+    };
+    messages.push(&final_message);
 
-    check_account_for_spend_multiple_fees_with_commitment(
+    check_payer(rpc_client, config, balance_needed, &messages)?;
+    send_deploy_messages(
         rpc_client,
-        &config.signers[0].pubkey(),
-        balance_needed,
-        &fee_calculator,
-        &messages,
-        config.commitment,
+        config,
+        &initial_message,
+        &write_messages,
+        &final_message,
+        buffer_signer,
+        program_signer,
     )?;
 
-    if let Some(message) = initial_message {
-        trace!("Creating or modifying program account");
-        let num_required_signatures = message.header.num_required_signatures;
+    Ok(json!({
+        "programId": format!("{}", program_signer.pubkey()),
+    })
+    .to_string())
+}
 
-        let mut initial_transaction = Transaction::new_unsigned(message);
-        // Most of the initial_transaction combinations require both the fee-payer and new program
-        // account to sign the transaction. One (transfer) only requires the fee-payer signature.
-        // This check is to ensure signing does not fail on a KeypairPubkeyMismatch error from an
-        // extraneous signature.
-        if num_required_signatures == 2 {
-            initial_transaction.try_sign(&signers, blockhash)?;
-        } else {
-            initial_transaction.try_sign(&[signers[0]], blockhash)?;
-        }
-        let result = rpc_client.send_and_confirm_transaction_with_spinner_and_config(
-            &initial_transaction,
-            config.commitment,
-            config.send_transaction_config,
+fn process_upgrade(
+    rpc_client: &RpcClient,
+    config: &CliConfig,
+    program_location: &str,
+    program: Pubkey,
+    upgrade_authority: SignerIndex,
+    buffer: Option<SignerIndex>,
+) -> ProcessResult {
+    const WORDS: usize = 12;
+    // Create ephemeral keypair to use for Buffer account, if not provided
+    let mnemonic = Mnemonic::new(MnemonicType::for_word_count(WORDS)?, Language::English);
+    let seed = Seed::new(&mnemonic, "");
+    let new_keypair = keypair_from_seed(seed.as_bytes())?;
```

Comment:
This is now duplicated between `process_deploy` and `progress_program_update()`, make a `fn`?